boa_gc/src/lib.rs

```diff
@@ -1,6 +1,575 @@
 //! Garbage collector for the Boa JavaScript engine.
 
-pub use gc::{
-    custom_trace, finalizer_safe, force_collect, unsafe_empty_trace, Finalize, Gc, GcCell as Cell,
-    GcCellRef as Ref, GcCellRefMut as RefMut, Trace,
-};
+#![allow(
+    clippy::let_unit_value,
+    clippy::should_implement_trait,
+    clippy::match_like_matches_macro,
+    clippy::new_ret_no_self,
+    // Putting the below on the allow list for now, but these should eventually be addressed
+    clippy::missing_safety_doc,
+    clippy::explicit_auto_deref,
+    clippy::borrow_deref_ref,
+)]
+
+use boa_profiler::Profiler;
+use std::cell::{Cell as StdCell, RefCell as StdRefCell};
+use std::mem;
+use std::ptr::NonNull;
+
+mod gc_box;
+mod internals;
+mod pointers;
+pub mod trace;
+
+pub use boa_gc_macros::{Finalize, Trace};
+
+pub use crate::trace::{Finalize, Trace};
+pub(crate) use gc_box::GcBox;
+pub use internals::{Ephemeron, GcCell as Cell, GcCellRef as Ref, GcCellRefMut as RefMut};
+pub use pointers::{Gc, WeakGc, WeakPair};
+
+pub type GcPointer = NonNull<GcBox<dyn Trace>>;
+
+// TODO: Determine if thread local variables are the correct approach vs an initialized structure
+thread_local!(pub static EPHEMERON_QUEUE: StdCell<Option<Vec<GcPointer>>> = StdCell::new(None));
+thread_local!(pub static GC_DROPPING: StdCell<bool> = StdCell::new(false));
+thread_local!(static BOA_GC: StdRefCell<BoaGc> = StdRefCell::new( BoaGc {
+    config: GcConfig::default(),
+    runtime: GcRuntimeData::default(),
+    adult_start: StdCell::new(None),
+    youth_start: StdCell::new(None),
+}));
```

Comment:
What do you mean by "initialized structure"? Like, putting everything on a single `thread_local!`?

Reply:
No, probably not the right way to word what I was thinking. I meant a non-thread local struct. Basically, I was thinking something that would have to be initialized by context when I wrote that comment.

Reply:
I think we can use thread locals for our std implementation, since they should be more performant on multi-threaded contexts (multiple Gcs can be run at the same time, because JS engines must always run on a single thread). On the other hand, we can use other mechanisms for no_std builds, because a system without an OS has no notion of threads.